### PR TITLE
HPCC-14522 Implement support for affinity in Thor

### PR DESCRIPTION
**Pull request by ghalliday**
Signed-off-by: Gavin Halliday gavin.halliday@lexisnexis.com

### CODE REVIEW

**Comment by ghalliday** [comment]
@jakesmith @mckellyln pushed for discussion.  We also need to make sure it compiles on all our target platforms.

**Comment by ghalliday** [comment]
@AttilaVamos the smoke test failed because a package numactl-devel would need to be installed.

**Comment by AttilaVamos** [comment]
@ghalliday installed.

**Comment by ghalliday** [comment]
@richardkchapman can you check if this compiles in osx?

@jakesmith @mckellyln please review.

**Comment by richardkchapman** [comment]
It does not - as far as I can determine there is no support for numa or numactl on OSX

**Comment by ghalliday** [comment]
ok.  I'll need to possibly make it conditional on numa being available, with a build option.
I'll wait until first review though.

File: cmake_modules/dependencies/el5.cmake [hunk 1]
```diff
@@ -1 +1 @@
-SET_DEPENDENCIES ( CPACK_RPM_PACKAGE_REQUIRES boost141-regex openldap libicu m4 libtool libxslt libxml2 gcc-c++ openssh-server openssh-clients expect libarchive rsync apr apr-util zip tbb )
+SET_DEPENDENCIES ( CPACK_RPM_PACKAGE_REQUIRES boost141-regex openldap libicu m4 libtool libxslt libxml2 gcc-c++ openssh-server openssh-clients expect libarchive rsync apr apr-util zip tbb numactl-devel )
```
Comment (mckellyln) [inline cmake_modules/dependencies/el5.cmake]: Probably dont need -devel package for installation, just numactl.  But do need it for building so its probably fine this way.  But inconsistent with Debian below where its not the -dev package listed.

Comment (richardkchapman) [inline cmake_modules/dependencies/el5.cmake]: These dependencies should list the run-time not compile-time dependencies

Comment (ghalliday) [inline cmake_modules/dependencies/el5.cmake]: I couldn't find a non development package that included the library - if there is one then please let me know what it is.

Comment (richardkchapman) [inline cmake_modules/dependencies/el5.cmake]: http://rpmfind.net/linux/rpm2html/search.php?query=numactl

File: system/jlib/jthread.cpp [hunk 2]
```diff
@@ -2335,3 +2336,104 @@ unsigned threadLogID()  // for use in logging
...
+    CPU_ZERO(&cpus);
+
+    const char * cur = cpuList;
+    loop
+    {
+        char * next;
+        unsigned cpu1 = getCpuId(cur, &next);
+        if (*next == '-')
+        {
+            const char * range = next+1;
```
Comment (mckellyln) [inline system/jlib/jthread.cpp]: range is unused and now fails with jlib warnings as errors.

File: system/jlib/jthread.cpp [hunk 3]
```diff
@@ -2335,3 +2336,104 @@ unsigned threadLogID()  // for use in logging
...
+    }
+
+    if (optNodes)
+        throw makeStringException(1, "Numa node list not yet supported");
+
+    unsigned numNumaNodes = numa_max_node()+1;
+    if (numNumaNodes <= 1)
+        return;
+
+    unsigned numNodes = 1;
```
Comment (mckellyln) [inline system/jlib/jthread.cpp]: numNodes is unused and now fails with jlib warnings as errors

File: system/jlib/jthread.cpp [hunk 4]
```diff
@@ -2335,3 +2336,104 @@ unsigned threadLogID()  // for use in logging
...
+        throw makeStringExceptionV(1, "Failed to set affinity for node %u", curNode);
+
+    DBGLOG("Process bound to numa node %u of %u", curNode, numNumaNodes);
+#endif
+}
+
+void bindMemoryToLocalNodes()
+{
+#if defined(LIBNUMA_API_VERSION) && (LIBNUMA_API_VERSION>=2)
+    numa_set_bind_policy(1);
```
Comment (mckellyln) [inline system/jlib/jthread.cpp]: Perhaps redundant, but why not also a call here to:
numa_set_localalloc()
just to ensure local node policy, especially if we are enforcing strict.

**Comment by richardkchapman** [comment]
@ghalliday  Back to you

**Comment by mckellyln** [comment]
For bindMemorytoLocalNodes() we could get the current cpu mask and use that to construct a nodemask and set the mem binding to that, as in for example:

void bindMemoryToLocalNodes()
{
#if defined(LIBNUMA_API_VERSION) && (LIBNUMA_API_VERSION>=2)
    numa_set_bind_policy(1);

    unsigned numNumaNodes = numa_max_node()+1;
    if (numNumaNodes <= 1)
        return;

    struct bitmask *cpus = numa_allocate_cpumask();
    bool ok = (numa_sched_getaffinity(0, cpus) != 0);
    if (ok)
    {
        struct bitmask *nodes = numa_allocate_nodemask();
        int numCpus = numa_num_configured_cpus();
        for (int i=0;i<numCpus;i++)
        {
            bool inSet = numa_bitmask_isbitset(cpus, i);
            if (inSet)
            {
                int node = numa_node_of_cpu(i);
                numa_bitmask_setbit(nodes, node);
            }
        }
        numa_set_membind(nodes);
        DBGLOG("Process memory bound to numa nodemask 0x%x (of %u nodes total)", (unsigned)(*(nodes->maskp)), numNumaNodes);
        numa_bitmask_free(nodes);
    }
    else
        DBGLOG("Failed to get affinity for memory binding");

    numa_bitmask_free(cpus);
#endif
}

**Comment by mckellyln** [comment]
Even more simple:

void bindMemoryToLocalNodes()
{
#if defined(LIBNUMA_API_VERSION) && (LIBNUMA_API_VERSION>=2)
    numa_set_bind_policy(1);

    unsigned numNumaNodes = numa_max_node()+1;
    if (numNumaNodes <= 1)
        return;
    struct bitmask *nodes = numa_get_run_node_mask();
    numa_set_membind(nodes);
    DBGLOG("Process memory bound to numa nodemask 0x%x (of %u nodes total)", (unsigned)(*(nodes->maskp)), numNumaNodes);
#endif
}

**Comment by richardkchapman** [comment]
@mckellyln Please re-review

**Comment by ghalliday** [comment]
I haven't yet looked at mark's memory binding suggestion....  Also we need to think about osx.

**Comment by richardkchapman** [comment]
Re OSX, we should just ignore the NUMA calls

**Comment by richardkchapman** [comment]
@ghalliday Needs rebasing for a clean merge

**Comment by ghalliday** [comment]
@richardkchapman now rebased.
I have checked it compiles on windows and linux.  Starting a 4 node Thor system now assigns two to socket 0 and two to socket 1 on my local system.

I implemented @mckellyln suggested change for the memory - but that options defaults off because I'm not convinced you want it enabled most of the time.

**Comment by ghalliday** [comment]
@richardkchapman

**Comment by mckellyln** [comment]
Good to merge

**Pull request merged**